From 20b679d300f8219f672127a60ff37b9476b1d752 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:38 -0400
Subject: [virt] kvm/vmx: Advance rip to after an ICEBP instruction

Message-id: <1406040016-3289-184-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86262
O-Subject: [RHEL7 PATCH v2 183/221] KVM: VMX: Advance rip to after an ICEBP instruction
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Huw Davies <huw@codeweavers.com>

When entering an exception after an ICEBP, the saved instruction
pointer should point to after the instruction.

This fixes the bug here: https://bugs.launchpad.net/qemu/+bug/1119686

Signed-off-by: Huw Davies <huw@codeweavers.com>
Reviewed-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit fd2a445a94d2ab6b39fb623dc02fee48d01a565a)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index eb3f2b1..8fb56e4 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -4841,6 +4841,9 @@ static int handle_exception(struct kvm_vcpu *vcpu)
         (KVM_GUESTDBG_SINGLESTEP | KVM_GUESTDBG_USE_HW_BP))) {
    vcpu->arch.dr6 &= ~15;
    vcpu->arch.dr6 |= dr6;
+   if (!(dr6 & ~DR6_RESERVED)) /* icebp */
+    skip_emulated_instruction(vcpu);
+
    kvm_queue_exception(vcpu, DB_VECTOR);
    return 1;
   }
-- 
1.7.1